File: benchmarks/tasks/binary_search/binary_search.py
# SPDX-License-Identifier: MIT-0

def binary_search(xs, target):
    lo, hi = 0, len(xs) - 1
    while lo <= hi:
        mid = (lo + hi) // 2
        if xs[mid] == target:
            return mid
        elif xs[mid] < target:
            lo = mid + 1
        else:
            hi = mid - 1
    return -1

print(binary_search([1, 2, 3, 4, 5, 6, 7, 8, 9, 10], 7))

File: benchmarks/tasks/binary_search/binary_search.hs
-- SPDX-License-Identifier: MIT-0

binarySearch :: [Int] -> Int -> Int
binarySearch xs target = go 0 (length xs - 1)
  where
    go lo hi
      | lo > hi        = -1
      | val == target  = mid
      | val < target   = go (mid + 1) hi
      | otherwise      = go lo (mid - 1)
      where
        mid = (lo + hi) `div` 2
        val = xs !! mid

main :: IO ()
main = print (binarySearch [1, 2, 3, 4, 5, 6, 7, 8, 9, 10] 7)
